aten/src/ATen/native/native_functions.yaml

```diff
@@ -105,6 +105,13 @@
   manual_kernel_registration: True
   variants: method
 
+- func: fw_primal(Tensor(a) self, int level) -> Tensor(a)
+  use_c10_dispatcher: full
+  manual_kernel_registration: True
+  variants: method
+  dispatch:
+    DefaultBackend: fw_primal
```

Comment:
The presence of the dispatch dictionary here is supremely perplexing, because if you *disabled* kernel registration, then dispatch dictionary doesn't do anything. So there is definitely something confused here.

Reply:
If I don't do that but provide a VariableType, then the codegen fails saying that this is missing.
I am definitely confused about why. But I followed what the error message said and then it succeeded.

Reply:
In particular, this is the error in question: https://github.com/pytorch/pytorch/blob/25833e5d1cba1ba3a8631d4014c37879337d00dc/tools/autograd/gen_variable_type.py#L380-L385

So you should actually set both if you want to do the registration manually no? (I agree it is not what we want here)

Reply:
The design tool I want you to use in this situation is this: what **should** the system do (as opposed to what "does" it do)? Error messages should be considered in terms of their *intent*, not the absolute specifics of how they are actually coded (which can be wrong).

In this particular case, you've hit an error message saying that you have a `derivatives.yaml` entry, you must also have a `dispatch` section. Why? If you omit the `dispatch` section from a kernel, that implies that you just want to register a "Math" kernel for the operator. Math kernels define their autograd in terms of the constituent operators they call, so it's an error if you also define something in `derivatives.yaml` since it will be ignored...

...except that when you have `manual_kernel_registration`, the `dispatch` table is invisible to `native_functions.yaml`, because you said that you wanted to do it manually. So what do you think should happen in this case?

Reply:
(UPDATE: It looks like in the most recent version of the patch, the kernels aren't manually registered anymore, which also solves this problem.)

Reply:
Yes thanks to your other comment, I looked into this more and realized that this should be a full function and only the VariableType implementation needs to be custom.

I also realized that I want this function to be internal and I need to update it to be marked as such!